File: Auto_stage_creator/c++/AutoStageCoordCreator.cpp
#include "AutoStageCoordCreator.h"
#include <fstream>
#include <sstream>
#include <iomanip>
#include <iostream>
#include <cmath>

AutoStageCoordCreator::AutoStageCoordCreator(double xmin, double xmax, 
                                           double ymin, double ymax, 
                                           double zmin, double zmax)
    : xmin(xmin), xmax(xmax), ymin(ymin), ymax(ymax), zmin(zmin), zmax(zmax),
      roundx(2), roundy(2), roundz(2) {
}

void AutoStageCoordCreator::setRounding(int x_round, int y_round, int z_round) {
    roundx = x_round;
    roundy = y_round;
    roundz = z_round;
}

void AutoStageCoordCreator::createCoordinates(const std::vector<double>& params) {
    if (params.size() != 9) {
        throw std::invalid_argument("Parameters array must contain exactly 9 elements");
    }
    
    double x_start = params[0];
    double y_start = params[1];
    double z_start = params[2];
    double x_end = params[3];
    double y_end = params[4];
    double z_end = params[5];
    int nx = static_cast<int>(params[6]);
    int ny = static_cast<int>(params[7]);
    int nz = static_cast<int>(params[8]);
    
    // Debug output to show what we received
    std::cout << "Received parameters:" << std::endl;
    std::cout << "X: " << x_start << " to " << x_end << " (" << nx << " steps)" << std::endl;
    std::cout << "Y: " << y_start << " to " << y_end << " (" << ny << " steps)" << std::endl;
    std::cout << "Z: " << z_start << " to " << z_end << " (" << nz << " steps)" << std::endl;
    
    // Apply bounds checking - only clamp if outside bounds (matching Python logic)
    // Note: The Python version has some inconsistent logic, but we'll implement proper clamping
    if (x_start < xmin) x_start = xmin;
    if (x_start > xmax) x_start = xmax;
    if (x_end < xmin) x_end = xmin;
    if (x_end > xmax) x_end = xmax;
    
    if (y_start < ymin) y_start = ymin;
    if (y_start > ymax) y_start = ymax;
    if (y_end < ymin) y_end = ymin;
    if (y_end > ymax) y_end = ymax;
    
    if (z_start < zmin) z_start = zmin;
    if (z_start > zmax) z_start = zmax;
    if (z_end < zmin) z_end = zmin;
    if (z_end > zmax) z_end = zmax;
    
    std::cout << "After bounds checking:" << std::endl;
    std::cout << "X: " << x_start << " to " << x_end << " (" << nx << " steps)" << std::endl;
    std::cout << "Y: " << y_start << " to " << y_end << " (" << ny << " steps)" << std::endl;
    std::cout << "Z: " << z_start << " to " << z_end << " (" << nz << " steps)" << std::endl;
    
    // Generate coordinate arrays
    std::vector<double> x_coords = linspace(x_start, x_end, nx);
    std::vector<double> y_coords = linspace(y_start, y_end, ny);
    std::vector<double> z_coords = linspace(z_start, z_end, nz);
    
    // Handle empty coordinate arrays
    if (x_coords.empty()) x_coords.push_back(x_start);
    if (y_coords.empty()) y_coords.push_back(y_start);
    if (z_coords.empty()) z_coords.push_back(z_start);
    
    std::cout << "Length of x_coords: " << x_coords.size() 
              << ", y_coords: " << y_coords.size() 
              << ", z_coords: " << z_coords.size() << std::endl;
    std::cout << "Will create " << x_coords.size() << " * " << y_coords.size() 
              << " * " << z_coords.size() << " = " 
              << (x_coords.size() * y_coords.size() * z_coords.size()) 
              << " coordinates" << std::endl;
    
    // Clear previous coordinates
    coordinates.clear();
    
    // Generate all coordinate combinations
    for (double x : x_coords) {
        for (double y : y_coords) {
            for (double z : z_coords) {
                coordinates.emplace_back(x, y, z);
            }
        }
    }
}

void AutoStageCoordCreator::writeCoordinates(const std::string& filename) {
    std::ofstream file(filename);
    if (!file.is_open()) {
        throw std::runtime_error("Could not open file for writing: " + filename);
    }
    
    if (coordinates.empty()) {
        throw std::runtime_error("No coordinates to write");
    }
    
    // Write first coordinate without newline prefix
    auto [x, y, z] = coordinates[0];
    file << formatDouble(x, roundx) << "," 
         << formatDouble(y, roundy) << "," 
         << formatDouble(z, roundz);
    
    // Write remaining coordinates with newline prefix
    for (size_t i = 1; i < coordinates.size(); ++i) {
        auto [x, y, z] = coordinates[i];
        file << "\n" << formatDouble(x, roundx) << "," 
             << formatDouble(y, roundy) << "," 
             << formatDouble(z, roundz);
    }
    
    file.close();
}

void AutoStageCoordCreator::clearCoordinates() {
    coordinates.clear();
}

std::vector<double> AutoStageCoordCreator::linspace(double start, double end, int num) {
    std::vector<double> result;
    
    if (num <= 0) {
        return result;
    }
    
    if (num == 1) {
        result.push_back(start);
        return result;
    }
    
    double step = (end - start) / (num - 1);
    for (int i = 0; i < num; ++i) {
        result.push_back(start + i * step);
    }
    
    return result;
}

std::string AutoStageCoordCreator::formatDouble(double value, int precision) {
    std::ostringstream oss;
    oss << std::fixed << std::setprecision(precision) << value;
    return oss.str();
}

File: Auto_stage_creator/c++/autostage_coord_creator.py

import numpy as np
import tkinter as tk
from tkinter import ttk, filedialog
class AutoStageCoordcreator:
    def __init__(self, xmin=0, xmax=300, ymin=0, ymax=300, zmin=0, zmax=300):
        self.xmin = xmin
        self.xmax = xmax
        self.ymin = ymin
        self.ymax = ymax
        self.zmin = zmin
        self.zmax = zmax
        self.makecoordinates = []
        self.roundx = 2
        self.roundy = 2
        self.roundz = 2

    def set_rounding(self, x_round=2, y_round=2, z_round=2):
        self.roundx = x_round
        self.roundy = y_round
        self.roundz = z_round
    
    def create_coordinates(self, arr):
        # arr consits of (x-start, y-start, z-start, x-end, y-end, z-end, N-x-steps, N-y-steps, N-z-steps)
        x_start, y_start, z_start, x_end, y_end, z_end, nx, ny, nz = arr

        if x_start > self.xmin:
            x_start = self.xmin
        if x_end < self.xmax:
            x_end = self.xmax
        if y_start > self.ymin:
            y_start = self.ymin
        if y_end < self.ymax:
            y_end = self.ymax
        if z_start > self.zmin:
            z_start = self.zmin
        elif z_end < self.zmax:
            z_end = self.zmax

        x_coords = np.linspace(x_start, x_end, nx)
        if len(x_coords) == 0:
            x_coords = [x_start]
        y_coords = np.linspace(y_start, y_end, ny)
        if len(y_coords) == 0:
            y_coords = [y_start]
        z_coords = np.linspace(z_start, z_end, nz)
        if len(z_coords) == 0:
            z_coords = [z_start]
        print(f"length of x_coords: {len(x_coords)}, y_coords: {len(y_coords)}, z_coords: {len(z_coords)}")
        print(f"will create {len(x_coords)} * {len(y_coords)} * {len(z_coords)} = {len(x_coords) * len(y_coords) * len(z_coords)} coordinates")

        for x in x_coords:
            for y in y_coords:
                for z in z_coords:
                    self.makecoordinates.append((f"{round(x, int(self.roundx)):.{self.roundx}f}", 
                                                 f"{round(y, int(self.roundy)):.{self.roundy}f}", 
                                                 f"{round(z, int(self.roundz)):.{self.roundz}f}"))
    
    def write_coordinates(self, filename):
        with open(filename, 'w') as f:
            f.write(f"{self.makecoordinates[0][0]},{self.makecoordinates[0][1]},{self.makecoordinates[0][2]}")
            for i in range(len(self.makecoordinates)-1):
                f.write(f"\n{self.makecoordinates[i+1][0]},{self.makecoordinates[i+1][1]},{self.makecoordinates[i+1][2]}")


class AutoStageGUI:
    def __init__(self):
        self.root = tk.Tk()
        self.root.title("AutoStage Coordinate Creator")
        self.asc = AutoStageCoordcreator()
        
        # Input fields
        frame = ttk.Frame(self.root, padding="10")
        frame.grid(row=0, column=0, sticky=(tk.W, tk.E, tk.N, tk.S))
        
        ttk.Label(frame, text="X Start:").grid(row=0, column=0)
        self.x_start = ttk.Entry(frame)
        self.x_start.grid(row=0, column=1)
        
        ttk.Label(frame, text="Y Start:").grid(row=1, column=0)
        self.y_start = ttk.Entry(frame)
        self.y_start.grid(row=1, column=1)
        
        ttk.Label(frame, text="Z Start:").grid(row=2, column=0)
        self.z_start = ttk.Entry(frame)
        self.z_start.grid(row=2, column=1)
        
        ttk.Label(frame, text="X End:").grid(row=0, column=2)
        self.x_end = ttk.Entry(frame)
        self.x_end.grid(row=0, column=3)
        
        ttk.Label(frame, text="Y End:").grid(row=1, column=2)
        self.y_end = ttk.Entry(frame)
        self.y_end.grid(row=1, column=3)
        
        ttk.Label(frame, text="Z End:").grid(row=2, column=2)
        self.z_end = ttk.Entry(frame)
        self.z_end.grid(row=2, column=3)
        
        ttk.Label(frame, text="X Steps:").grid(row=3, column=0)
        self.nx = ttk.Entry(frame)
        self.nx.grid(row=3, column=1)
        
        ttk.Label(frame, text="Y Steps:").grid(row=3, column=2)
        self.ny = ttk.Entry(frame)
        self.ny.grid(row=3, column=3)
        
        ttk.Label(frame, text="Z Steps:").grid(row=4, column=0)
        self.nz = ttk.Entry(frame)
        self.nz.grid(row=4, column=1)
        
        ttk.Button(frame, text="Create Coordinates", command=self.create_coords).grid(row=5, column=1)
        ttk.Button(frame, text="Save to File", command=self.save_file).grid(row=5, column=2)

    def create_coords(self):
        try:
            coords = [
                float(self.x_start.get()), float(self.y_start.get()),
                float(self.z_start.get()), float(self.x_end.get()),
                float(self.y_end.get()), float(self.z_end.get()),
                int(self.nx.get()), int(self.ny.get()), int(self.nz.get())
            ]
            self.asc.create_coordinates(coords)
        except ValueError:
            tk.messagebox.showerror("Error", "Invalid input values")

    def save_file(self):
        filename = filedialog.asksaveasfilename(defaultextension=".txt")
        if filename:
            self.asc.write_coordinates(filename)

    def run(self):
        self.root.mainloop()

def asctest():
    ASC = AutoStageCoordcreator()
    ASC.set_rounding(2, 2, 2)
    ASC.create_coordinates([121, 101.3, 150, 126.32, 132.23, 150, 43, 86, 0])
    textfile = 'c2.txt'
    ASC.write_coordinates(textfile)
    print(f"Coordinates created and written to {textfile}.")

def rungui():
    gui = AutoStageGUI()
    gui.run()

# example usage
if __name__ == "__main__":
    rungui()
    # asctest()  # Uncomment to run the test function instead of the GUI
